File: scripts/nats/jetstream_storage_policy.py
#!/usr/bin/env python3
"""Declared stream capacity shared by broker generation and deploy admission.

This enforces allocation parity, not measured throughput or outage retention.
SENSOR-CRITICAL-108 retains ownership of those measurements and sizing decisions.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

STORAGE_POLICY = (
    Path(__file__).resolve().parents[2]
    / "platform/libs/event-bus/src/nats/jetstream-storage-policy.json"
)
MAX_SAFE_INTEGER = 9007199254740991


def positive_safe_integer(value: Any) -> bool:
    return type(value) is int and 0 < value <= MAX_SAFE_INTEGER


def load_storage_policy() -> dict[str, Any]:
    policy = json.loads(STORAGE_POLICY.read_text())
    if (
        not isinstance(policy, dict)
        or set(policy) != {"schema_version", "reserve", "streams"}
        or type(policy["schema_version"]) is not int
        or policy["schema_version"] != 1
        or not isinstance(policy["reserve"], dict)
        or set(policy["reserve"]) != {"numerator", "denominator"}
        or not isinstance(policy["streams"], dict)
        or set(policy["streams"]) != {"events", "telemetry", "dlq"}
    ):
        raise ValueError("invalid canonical JetStream storage policy shape")
    reserve = policy["reserve"]
    if (
        not positive_safe_integer(reserve["numerator"])
        or not positive_safe_integer(reserve["denominator"])
        or reserve["numerator"] < reserve["denominator"]
    ):
        raise ValueError("invalid canonical JetStream storage reserve")
    for stream in policy["streams"].values():
        if (
            not isinstance(stream, dict)
            or set(stream) != {"max_bytes"}
            or not positive_safe_integer(stream["max_bytes"])
        ):
            raise ValueError("invalid canonical JetStream stream allocation")
    return policy


def required_file_store_bytes(policy: dict[str, Any], override: str = "") -> int:
    reserved = sum(stream["max_bytes"] for stream in policy["streams"].values())
    reserved *= policy["reserve"]["numerator"]
    if not positive_safe_integer(reserved):
        raise ValueError("canonical JetStream reservation exceeds safe integer arithmetic")
    denominator = policy["reserve"]["denominator"]
    required = (reserved + denominator - 1) // denominator
    if override:
        if not re.fullmatch(r"[0-9]+", override) or not positive_safe_integer(int(override)):
            raise ValueError("NATS_REQUIRED_FILE_STORE_BYTES must be a positive safe integer")
        # Operator policy may require more capacity, never less than runtime.
        required = max(required, int(override))
    return required


def main() -> int:
    if len(sys.argv) not in (2, 3) or sys.argv[1] != "--required-file-store":
        sys.stderr.write("usage: jetstream_storage_policy.py --required-file-store [operator-floor]\n")
        return 64
    try:
        policy = load_storage_policy()
        override = sys.argv[2] if len(sys.argv) == 3 else ""
        print(required_file_store_bytes(policy, override))
    except (OSError, ValueError) as error:
        sys.stderr.write(f"error: JetStream capacity policy: {error}\n")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: platform/libs/event-bus/src/nats/jetstream-storage-policy.ts
import policy from './jetstream-storage-policy.json';

type JetStreamStorageKind = keyof typeof policy.streams;

function requiredFileStoreBytes(): number {
  const allocations = Object.values(policy.streams).map((stream) => stream.max_bytes);
  const { numerator, denominator } = policy.reserve;
  if (
    policy.schema_version !== 1 ||
    !allocations.every((bytes) => Number.isSafeInteger(bytes) && bytes > 0) ||
    !Number.isSafeInteger(numerator) ||
    !Number.isSafeInteger(denominator) ||
    denominator < 1 ||
    numerator < denominator
  ) {
    throw new Error('Invalid canonical JetStream storage policy');
  }
  const reservedBytes = allocations.reduce((total, bytes) => total + bytes, 0) * numerator;
  if (!Number.isSafeInteger(reservedBytes)) {
    throw new Error('Canonical JetStream storage reservation exceeds safe integer arithmetic');
  }
  return Math.ceil(reservedBytes / denominator);
}

export const JETSTREAM_REQUIRED_FILE_STORE_BYTES = requiredFileStoreBytes();

/** Environment overrides may reduce a stream allocation, but never exceed the
 * capacity the same immutable policy reserves in the broker and deploy gate. */
export function getJetStreamStorageBudget(
  stream: JetStreamStorageKind,
  configuredValue?: unknown,
): number {
  const declared = policy.streams[stream].max_bytes;
  if (configuredValue === undefined) return declared;
  const parsed =
    typeof configuredValue === 'number'
      ? configuredValue
      : typeof configuredValue === 'string' && /^[0-9]+$/.test(configuredValue)
        ? Number(configuredValue)
        : Number.NaN;
  if (!Number.isSafeInteger(parsed) || parsed < 1 || parsed > declared) {
    throw new Error(
      `NATS_${stream.toUpperCase()}_MAX_BYTES must be a positive safe integer no greater than its declared allocation (${declared})`,
    );
  }
  return parsed;
}
